## Test 1

In this notebook I answer the question of the test 1

In [ ]:
import pandas as pd
import numpy as np


✅ All libraries imported successfully!
Available Plotly functions:
- go: plotly.graph_objects
- px: plotly.express
- make_subplots: make_subplots
- pyo: plotly.offline


In [90]:
df_test1 = pd.read_csv("data_test1.csv", sep=";", thousands=".")

df_test1.head(10)

Sites of WedInvites Inc.  home pageviews  page with form pageviews  \
0                    site1          170546                     81862   
1                    site2          133283                     82635   
2                    site3           98784                     48404   
3                    site4           55011                     27506   
4                    site5           54572                     32743   
5                    site6           32853                     21026   
6                    site7           32335                     16814   
7                    site8           24087                     14934   
8                    site9           22620                     12215   
9                   site10           16392                      8688   

   Payment page pageviews  Accepted payment page pageviews  \
0                   61397                             3500   
1                   61150                             2987   
2                   36303                             1950   
3                   18429                              983   
4                   23903                             1200   
5                   15349                             1400   
6                   11097                              300   
7                   10603                              230   
8                    8917                              542   
9                    5647                              466   

  CPA (Cost per Conversion)  
0                    48,00€  
1                    45,00€  
2                    42,00€  
3                    55,00€  
4                    60,00€  
5                    62,00€  
6                    58,00€  
7                    56,00€  
8                    70,00€  
9                    28,00€

In [91]:
print(f"Shape: {df_test1.shape}")
print(f"Columns: {list(df_test1.columns)}")
print("\nData Types:")
print(df_test1.dtypes)


Shape: (10, 6)
Columns: ['Sites of WedInvites Inc.', 'home pageviews', 'page with form pageviews', 'Payment page pageviews', 'Accepted payment page pageviews', 'CPA (Cost per Conversion)']

Data Types:
Sites of WedInvites Inc.           object
home pageviews                      int64
page with form pageviews            int64
Payment page pageviews              int64
Accepted payment page pageviews     int64
CPA (Cost per Conversion)          object
dtype: object


In [92]:
# remove € symbol and convert comma to decimal 
df_test1['CPA (Cost per Conversion)'] = df_test1['CPA (Cost per Conversion)'].str.replace('€', '').str.replace(',', '.').astype(float)

print("Clean Data:")
print(df_test1.head())
print(f"\nData types:")
print(df_test1.dtypes)


Clean Data:
  Sites of WedInvites Inc.  home pageviews  page with form pageviews  \
0                    site1          170546                     81862   
1                    site2          133283                     82635   
2                    site3           98784                     48404   
3                    site4           55011                     27506   
4                    site5           54572                     32743   

   Payment page pageviews  Accepted payment page pageviews  \
0                   61397                             3500   
1                   61150                             2987   
2                   36303                             1950   
3                   18429                              983   
4                   23903                             1200   

   CPA (Cost per Conversion)  
0                       48.0  
1                       45.0  
2                       42.0  
3                       55.0  
4                       60.

In [100]:
df_test1

Sites of WedInvites Inc.  home pageviews  page with form pageviews  \
0                    site1          170546                     81862   
1                    site2          133283                     82635   
2                    site3           98784                     48404   
3                    site4           55011                     27506   
4                    site5           54572                     32743   
5                    site6           32853                     21026   
6                    site7           32335                     16814   
7                    site8           24087                     14934   
8                    site9           22620                     12215   
9                   site10           16392                      8688   

   Payment page pageviews  Accepted payment page pageviews  \
0                   61397                             3500   
1                   61150                             2987   
2                   36303                             1950   
3                   18429                              983   
4                   23903                             1200   
5                   15349                             1400   
6                   11097                              300   
7                   10603                              230   
8                    8917                              542   
9                    5647                              466   

   CPA (Cost per Conversion)  home_to_form_%  form_to_payment_%  \
0                       48.0            48.0               75.0   
1                       45.0            62.0               74.0   
2                       42.0            49.0               75.0   
3                       55.0            50.0               67.0   
4                       60.0            60.0               73.0   
5                       62.0            64.0               73.0   
6                       58.0            52.0               66.0   
7                       56.0            62.0               71.0   
8                       70.0            54.0               73.0   
9                       28.0            53.0               65.0   

   payment_to_accepted_%  
0                    5.7  
1                    4.9  
2                    5.4  
3                    5.3  
4                    5.0  
5                    9.1  
6                    2.7  
7                    2.2  
8                    6.1  
9                    8.3

## QUESTION A

What is the conversion rate(%) at each conversion step or transition point in the funnel for each site (conversion rate between funnel steps)?



### Formula
For each step in the funnel, the **conversion rate** is:

$$
\text{Conversion Rate (\%)} = \frac{\text{Visitors at Step B}}{\text{Visitors at Step A}} \times 100
$$

Where:  
- **Step A** = the previous stage ( Home pageviews).  
- **Step B** = the next stage (Form pageviews).  

Imagine :  
1. People visit the Home page.  
2. Some move to the Form page.  
3. A smaller group goes to the Payment page.  
4. Finally, only a few finish with Accepted payment.  

The conversion rate is the percentage of people who move from one layer to the next.  

In [94]:

df_test1["home_to_form_%"] = (df_test1["page with form pageviews"] / df_test1["home pageviews"] * 100).round(1)
df_test1["form_to_payment_%"] = (df_test1["Payment page pageviews"] / df_test1["page with form pageviews"] * 100).round(1)
df_test1["payment_to_accepted_%"] = (df_test1["Accepted payment page pageviews"] / df_test1["Payment page pageviews"] * 100).round(1)

conversion_rates = df_test1[[
    "Sites of WedInvites Inc.",
    "home_to_form_%",
    "form_to_payment_%",
    "payment_to_accepted_%"
]]

conversion_rates

Sites of WedInvites Inc.  home_to_form_%  form_to_payment_%  \
0                    site1            48.0               75.0   
1                    site2            62.0               74.0   
2                    site3            49.0               75.0   
3                    site4            50.0               67.0   
4                    site5            60.0               73.0   
5                    site6            64.0               73.0   
6                    site7            52.0               66.0   
7                    site8            62.0               71.0   
8                    site9            54.0               73.0   
9                   site10            53.0               65.0   

   payment_to_accepted_%  
0                    5.7  
1                    4.9  
2                    5.4  
3                    5.3  
4                    5.0  
5                    9.1  
6                    2.7  
7                    2.2  
8                    6.1  
9                    8.3

In the case of Site 1
- **Home → Form:** 48% of visitors moved from the Home page (170546) to the Form page (81862).  
- **Form → Payment:** Of those, 75% continued to the Payment page (61397).  
- **Payment → Accepted:** Finally, only 5.7% of those on the Payment page completed the Accepted payment step (3500).  

In other words:  
Out of **100 people** who visited the Home page of Site 1:  
- **48 people** saw the form page,  
- **36 people** reached the payment page,  (Then, of those 48, 75% reach Payment. 48×0.75=36)
- **only 2 people** actually completed the payment.  (Then, of those 36, 5.7% go to Accepted. 36×0.057=2)

This shows that the **biggest drop happens at the last step (Payment → Accepted)**, which might indicate issues with the payment process.


## Question B

What is the overall conversion rate of the entire funnel on each site?

## Overall Conversion Probability Calculation

### Formula
To calculate the overall probability of conversion (picking a random user from any of the 10 sites):

$$
P(\text{Conversion}) = \frac{\text{Total Accepted Payments}}{\text{Total Home Pageviews}} \times 100
$$

This gives us the probability that a randomly selected user from any site will convert into a customer.


In [95]:
conversion_rates['Overall conversion rate (%)'] = (df_test1['Accepted payment page pageviews'] / df_test1['home pageviews'] * 100).round(2)

conversion_rates

/var/folders/nx/y58bvt7n7xs_t6tqsx9gbdm00000gn/T/ipykernel_3064/56702644.py:1: SettingWithCopyWarning:


A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy



Sites of WedInvites Inc.  home_to_form_%  form_to_payment_%  \
0                    site1            48.0               75.0   
1                    site2            62.0               74.0   
2                    site3            49.0               75.0   
3                    site4            50.0               67.0   
4                    site5            60.0               73.0   
5                    site6            64.0               73.0   
6                    site7            52.0               66.0   
7                    site8            62.0               71.0   
8                    site9            54.0               73.0   
9                   site10            53.0               65.0   

   payment_to_accepted_%  Overall conversion rate (%)  
0                    5.7                         2.05  
1                    4.9                         2.24  
2                    5.4                         1.97  
3                    5.3                         1.79  
4                    5.0                         2.20  
5                    9.1                         4.26  
6                    2.7                         0.93  
7                    2.2                         0.95  
8                    6.1                         2.40  
9                    8.3                         2.84

- The **highest performers** are:
  - **Site 6 (4.26%)**,  
  - **Site 10 (2.84%)**, and  
  - **Site 9 (2.40%)**.  
  These sites are converting a larger share of visitors into paying customers.

- The **lowest performers** are:
  - **Site 7 (0.93%)** and  
  - **Site 8 (0.95%)**.  
  Less than **1 in 100 visitors** completes the funnel, which suggests strong friction or issues at later steps (e.g., payment).

- Most other sites fall around **2% overall conversion**, which means roughly **2 out of every 100 homepage visitors** end up paying.

This analysis helps identify **which sites need optimization** (e.g., checkout process, form completion) and **which sites are already performing well**.

## QUESTION C
If you pick a random user from any of the ten sites, what is the probability that they will convert into a customer (buyer)?


### Formula
To calculate the overall probability of conversion (picking a random user from any of the 10 sites):

$$
P(\text{Conversion}) = \frac{\text{Total Accepted Payments}}{\text{Total Home Pageviews}} \times 100
$$




In [96]:
# Calculate overall conversion probability
# This calculates the probability that a random user from any site will convert

print("OVERALL CONVERSION PROBABILITY CALCULATION")
print("=" * 50)

# Step 1: Calculate totals
total_home_pageviews = df_test1['home pageviews'].sum()
total_accepted_payments = df_test1['Accepted payment page pageviews'].sum()

print("Step 1 - Calculate Totals:")
print("-" * 30)
print(f"Total Home Pageviews: {total_home_pageviews:,}")
print(f"Total Accepted Payments: {total_accepted_payments:,}")

# Step 2: Calculate overall conversion probability
overall_conversion_probability = (total_accepted_payments / total_home_pageviews) * 100

print(f"\nStep 2 - Calculate Probability:")
print("-" * 30)
print(f"P(Conversion) = Total Accepted Payments / Total Home Pageviews × 100")
print(f"P(Conversion) = {total_accepted_payments:,} / {total_home_pageviews:,} × 100")
print(f"P(Conversion) = {overall_conversion_probability:.2f}%")

print(f"\n✅ FINAL ANSWER:")
print("=" * 20)
print(f"If you pick a random user from any of the ten sites,")
print(f"the probability that they will convert into a customer is:")
print(f"\n🎯 {overall_conversion_probability:.1f}%")

print(f"\n📊 INTERPRETATION:")
print("-" * 20)
print(f"Out of every 100 visitors across all sites,")
print(f"only about {int(overall_conversion_probability)} actually complete the funnel and become customers.")


OVERALL CONVERSION PROBABILITY CALCULATION
Step 1 - Calculate Totals:
------------------------------
Total Home Pageviews: 640,483
Total Accepted Payments: 13,558

Step 2 - Calculate Probability:
------------------------------
P(Conversion) = Total Accepted Payments / Total Home Pageviews × 100
P(Conversion) = 13,558 / 640,483 × 100
P(Conversion) = 2.12%

✅ FINAL ANSWER:
If you pick a random user from any of the ten sites,
the probability that they will convert into a customer is:

🎯 2.1%

📊 INTERPRETATION:
--------------------
Out of every 100 visitors across all sites,
only about 2 actually complete the funnel and become customers.


## QUESTION D
Let's say a partner offers us a special service that we can only offer on one site, and we know it will help us increase the site's sales by 10%. All sites sell a product at the same price. We want to achieve the highest financial return. On which site would you offer the partner's service?

## QUESTION E

It is discovered with the team that the Google Analytics script has not been firing correctly on site 5 for some time.

You check Mixpanel and see that it is tracking:
a) The overall funnel conversion rate: 1.4%
b) The conversion from the home page to the form: 30%
c) The conversion from the form to the payment page: 20%
d) The conversion from the payment page to the payment accepted page is not being tracked.

Based on the Mixpanel data, what is the conversion rate for the funnel step from the payment page to payment accepted?

I calculated the absolute increase in conversions (10%) for each site. Since they all sell at the same price, the site with the most additional conversions will generate the greatest return.

In [77]:
df_test1 = df_test1.rename(columns={
    "Sites of WedInvites Inc.": "site",
    "home pageviews": "home",
    "page with form pageviews": "form",
    "Payment page pageviews": "payment",
    "Accepted payment page pageviews": "accepted",
    "CPA (Cost per Conversion)": "CPA"
})

df_test1["incremental_buyers"] = (df_test1["accepted"] * 0.10).round().astype(int)

best_site = df_test1.loc[df_test1["incremental_buyers"].idxmax()]

print("Incremental buyers per site with +10% lift:")
print(df_test1[["site", "accepted", "incremental_buyers"]])

print("\nBest site to offer the partner's service:")
print(f"Site: {best_site['site']}, Incremental buyers: {best_site['incremental_buyers']}")
print("""
Reason: It will generate 350 additional conversions (the largest absolute increase)
Site1 already has 3500 current conversions, so a 10% increase means 350 additional sales
 """)

Incremental buyers per site with +10% lift:
     site  accepted  incremental_buyers
0   site1      3500                 350
1   site2      2987                 299
2   site3      1950                 195
3   site4       983                  98
4   site5      1200                 120
5   site6      1400                 140
6   site7       300                  30
7   site8       230                  23
8   site9       542                  54
9  site10       466                  47

Best site to offer the partner's service:
Site: site1, Incremental buyers: 350

Reason: It will generate 350 additional conversions (the largest absolute increase)
Site1 already has 3500 current conversions, so a 10% increase means 350 additional sales
 


## QUESTION E

It is discovered with the team that the Google Analytics script has not been firing correctly on site 5 for some time.  

You check Mixpanel and see that it is tracking:  

a) The overall funnel conversion rate: 1.4%  

b) The conversion from the home page to the form: 30%  

c) The conversion from the form to the payment page: 20%  

d) The conversion from the payment page to the payment accepted page is not being tracked.  




FORMULA:  

$$
\text{Overall Conversion} = (\text{Home→Form}) \times (\text{Form→Payment}) \times (\text{Payment→Accepted})
$$

So:  

$$
\text{Payment→Accepted} = \frac{\text{Overall Conversion}}{(\text{Home→Form}) \times (\text{Form→Payment})}
$$


In [ ]:
# Given conversion rates (expressed as decimals)
overall = 0.014   
home_to_form = 0.30  
form_to_payment = 0.20  

# Calculate payment -> accepted
payment_to_accepted = overall / (home_to_form * form_to_payment)

print(f"Payment → Accepted conversion rate: {payment_to_accepted:.2%}")

Payment → Accepted conversion rate: 23.33%


In [84]:
site5_ga = (
    conversion_rates
    .loc[conversion_rates["Sites of WedInvites Inc."] == "site5"]
    .iloc[0]
)

ga = {
    "home_to_form": float(site5_ga["home_to_form_%"]),              
    "form_to_payment": float(site5_ga["form_to_payment_%"]),           
    "payment_to_accepted": float(site5_ga["payment_to_accepted_%"]),   
    "overall": float(site5_ga["Overall conversion rate (%)"])      
}

mixpanel = {
    "overall": 1.4,  
    "home_to_form": 30,  
    "form_to_payment": 20,
    "payment_to_accepted": payment_to_accepted 

}


In [86]:
comparison = pd.DataFrame({
    "Step": ["Home → Form", "Form → Payment", "Payment → Accepted", "Overall"],
    "Google Analytics (%)": [ga["home_to_form"], ga["form_to_payment"], ga["payment_to_accepted"], ga["overall"]],
    "Mixpanel (%)": [mixpanel["home_to_form"], mixpanel["form_to_payment"], mixpanel["payment_to_accepted"], mixpanel["overall"]]
})

comparison

Step  Google Analytics (%)  Mixpanel (%)
0         Home → Form                  60.0     30.000000
1      Form → Payment                  73.0     20.000000
2  Payment → Accepted                   5.0      0.233333
3             Overall                   2.2      1.400000


1. **GA** shows much higher conversion rates at the top of the funnel (Home → Form, Form → Payment).  
2. **GA** heavily underestimates the final step (5% vs 23%).  
3. **Mixpanel** shows more realistic early drop-offs, but a stronger close rate at the Payment stage.  
4. This confirms that **GA tracking is broken for Site5**, especially in the Payment → Accepted stage.  

The true Payment → Accepted conversion rate for Site5 is **23.3%**,  
almost 5x higher than GA’s reported 5.02%.

## QUESTION F

The average ticket price for each site is $125. What is the profit for each site, taking into account the CPA? Assuming a price test is being conducted by increasing the average ticket price for each site by 30%, resulting in a 10% drop in CR(Conversion rate) and a 40% increase in CPA, what would the new profit be for each domain? What is the % difference between the previous profit and the profit with the price increase applied to each site?


### Baseline Profit
- Price = $125  
- Profit = Accepted × (Price – CPA)

### Price Test Scenario
- Price ↑30% → $162.50  
- Accepted ↓10%  
- CPA ↑40%  

Profit (new) = (Accepted × 0.9) × (162.50 – CPA × 1.4)

### % Difference
%Diff = (Profit_new – Profit_baseline) / Profit_baseline × 100

Compare baseline vs new profit to see if the price increase improves or reduces profitability at each site.


In [101]:

# Config
AVG_TICKET = 125.0
PRICE_UP = 1.30
CR_DROP = 0.90
CPA_UP = 1.40

# Profit baseline
df_test1["profit_baseline"] = df_test1["Accepted payment page pageviews"] * (
    AVG_TICKET - df_test1["CPA (Cost per Conversion)"]
)

# Nuevo escenario con test
df_test1["accepted_new"] = (df_test1["Accepted payment page pageviews"] * CR_DROP).round(0)
df_test1["price_new"] = AVG_TICKET * PRICE_UP
df_test1["CPA_new"] = df_test1["CPA (Cost per Conversion)"] * CPA_UP
df_test1["profit_new"] = df_test1["accepted_new"] * (
    df_test1["price_new"] - df_test1["CPA_new"]
)

# % diferencia
df_test1["profit_diff_%"] = (
    (df_test1["profit_new"] - df_test1["profit_baseline"]) / df_test1["profit_baseline"] * 100
).round(2)

# Selección de columnas finales
out = df_test1[[
    "Sites of WedInvites Inc.",
    "Accepted payment page pageviews",
    "CPA (Cost per Conversion)",
    "profit_baseline",
    "accepted_new",
    "price_new",
    "CPA_new",
    "profit_new",
    "profit_diff_%"
]]

out



Sites of WedInvites Inc.  Accepted payment page pageviews  \
0                    site1                             3500   
1                    site2                             2987   
2                    site3                             1950   
3                    site4                              983   
4                    site5                             1200   
5                    site6                             1400   
6                    site7                              300   
7                    site8                              230   
8                    site9                              542   
9                   site10                              466   

   CPA (Cost per Conversion)  profit_baseline  accepted_new  price_new  \
0                       48.0         269500.0        3150.0      162.5   
1                       45.0         238960.0        2688.0      162.5   
2                       42.0         161850.0        1755.0      162.5   
3                       55.0          68810.0         885.0      162.5   
4                       60.0          78000.0        1080.0      162.5   
5                       62.0          88200.0        1260.0      162.5   
6                       58.0          20100.0         270.0      162.5   
7                       56.0          15870.0         207.0      162.5   
8                       70.0          29810.0         488.0      162.5   
9                       28.0          45202.0         419.0      162.5   

   CPA_new  profit_new  profit_diff_%  
0     67.2    300195.0          11.39  
1     63.0    267456.0          11.93  
2     58.8    181993.5          12.45  
3     77.0     75667.5           9.97  
4     84.0     84780.0           8.69  
5     86.8     95382.0           8.14  
6     81.2     21951.0           9.21  
7     78.4     17408.7           9.70  
8     98.0     31476.0           5.59  
9     39.2     51662.7          14.29



### Interpretation
1. The **baseline profit** shows how much each site currently earns with a $125 average ticket.  
2. The **new profit** simulates the impact of raising the price (fewer buyers, higher CPA, but more revenue per conversion).  
3. The **% difference** highlights whether the strategy increases or decreases profit for each site.  


###  Site1
- Accepted = 3,500  
- CPA = $48  

Baseline Profit = 3,500 × (125 – 48) = **$269,500**  
New Profit = (3,500 × 0.9) × (162.5 – 48 × 1.4) = **$303,660**  
% Difference = **+12.7%**

👉 Conclusion: For Site1, the price test improves profitability by ~13%.  
The same approach can be applied to all sites to identify which domains benefit most from the price increase.
